<a href="https://colab.research.google.com/github/mfvalle/AlgorithmsUN2024I/blob/main/LAB4/mvalleGroup21Lab4.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
!cat /proc/cpuinfo

processor	: 0
vendor_id	: GenuineIntel
cpu family	: 6
model		: 79
model name	: Intel(R) Xeon(R) CPU @ 2.20GHz
stepping	: 0
microcode	: 0xffffffff
cpu MHz		: 2199.998
cache size	: 56320 KB
physical id	: 0
siblings	: 2
core id		: 0
cpu cores	: 1
apicid		: 0
initial apicid	: 0
fpu		: yes
fpu_exception	: yes
cpuid level	: 13
wp		: yes
flags		: fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ss ht syscall nx pdpe1gb rdtscp lm constant_tsc rep_good nopl xtopology nonstop_tsc cpuid tsc_known_freq pni pclmulqdq ssse3 fma cx16 pcid sse4_1 sse4_2 x2apic movbe popcnt aes xsave avx f16c rdrand hypervisor lahf_lm abm 3dnowprefetch invpcid_single ssbd ibrs ibpb stibp fsgsbase tsc_adjust bmi1 hle avx2 smep bmi2 erms invpcid rtm rdseed adx smap xsaveopt arat md_clear arch_capabilities
bugs		: cpu_meltdown spectre_v1 spectre_v2 spec_store_bypass l1tf mds swapgs taa mmio_stale_data retbleed
bogomips	: 4399.99
clflush size	: 64
cache_alignment	: 64
addres

# I Solving the Sum of Two Digits Programming Challenge

## C++

In [5]:
%%writefile APlusB.cpp
#include <iostream>

int sum_of_two_digits(int first_digit, int second_digit) {
    return first_digit + second_digit;
}

int main() {
    int a = 0;
    int b = 0;
    std::cin >> a;
    std::cin >> b;
    std::cout << sum_of_two_digits(a, b);
    return 0;
}

Writing APlusB.cpp


In [6]:
!g++ APlusB.cpp -o APlusB -std=c++11
!./APlusB

4 8
12

### Test input output files

In [7]:
%%writefile input1.txt
4 8

Writing input1.txt


In [8]:
%%writefile output1.txt
12

Writing output1.txt


In [9]:
%%writefile input2.txt
0 9

Writing input2.txt


In [10]:
%%writefile output2.txt
9

Writing output2.txt


In [11]:
!./APlusB < input1.txt

12

In [12]:
!./APlusB < input2.txt

9

### Comparing files with grep

https://askubuntu.com/questions/546796/comparing-two-text-files

In [13]:
!grep -xvFf output1.txt output1.txt

In [14]:
!grep -xvFf output1.txt output2.txt

9


In [15]:
!grep -xvFf output2.txt output1.txt

12


In [16]:
!./APlusB < input1.txt> ctest1.txt
!grep -xvFf output1.txt ctest1.txt

In [17]:
!./APlusB < input2.txt > ctest2.txt
!grep -xvFf output2.txt ctest2.txt

## Python

In [19]:
%%writefile sum.py
# python3

def sum_of_two_digits(first_digit, second_digit):
    return first_digit + second_digit

if __name__ == '__main__':
    a, b = map(int, input().split())
    print(sum_of_two_digits(a, b))

Writing sum.py


In [20]:
!python sum.py

4 8
12


In [21]:
!python sum.py  < input1.txt> ptest1.txt
!grep -xvFf output1.txt ptest1.txt

In [22]:
!python sum.py  < input2.txt> ptest2.txt
!grep -xvFf output2.txt ptest2.txt

# II Solving The Maximum Pairwise Product Programming Challenge in C++

## C++

$$ T(n) < 6 n^2 = O(n^2)$$

In [23]:
%%writefile max_pairwise_product_full.cpp
#include <iostream>
#include <vector>
#include <algorithm>

int MaxPairwiseProduct(const std::vector<int>& numbers) {
    int max_product = 0;
    int aux_product = 0;
    int n = numbers.size();

    for (int first = 0; first < n; ++first) {
        for (int second = 0; second < n; ++second) {
          if ( first != second ){
                aux_product = numbers[first] * numbers[second];
                if (aux_product > max_product)  max_product = aux_product;
          }
        }
    }

    return max_product;
}

int main() {
    int n;
    std::cin >> n;
    std::vector<int> numbers(n);
    for (int i = 0; i < n; ++i) {
        std::cin >> numbers[i];
    }

    std::cout << MaxPairwiseProduct(numbers) << "\n";
    return 0;
}

Writing max_pairwise_product_full.cpp


#### Time Complexity
$$\left. \begin{array}{llc}
 \text{INSTRUCTION} & \text{TYPE}& \text{# EXECUTIONS}\\
\text{first != second} &  \text{comparisons} &n^2 \\
\text{numbers[first] * numbers[second]}& \ \text{ products}  &n^2 - n \\
\text{max_product = aux_product} &  \text{assigments} & [1, \frac{n^2 - n}{2})  
\end{array}\right\}  =  O(n^2)$$

In [24]:
!g++ max_pairwise_product_full.cpp -o max_pairwise_product_full
!./max_pairwise_product_full

5
80 89 12 899 1200 200
1078800


### Test 1: Small numbers

In [25]:
%%writefile input0.txt
10
0 0 0 0 0 0 0 0 0 0

Writing input0.txt


In [26]:
%%writefile output0.txt
0

Writing output0.txt


In [27]:
!./max_pairwise_product_full< input0.txt > ctest0.txt
!grep -xvFf output0.txt ctest0.txt

In [30]:
%%writefile input1.txt
10
1 2 3 4 5 6 7 8 9 10

Overwriting input1.txt


In [33]:
!./max_pairwise_product_full < input1.txt

90


In [ ]:
%%writefile output1.txt
90

Overwriting output1.txt


In [34]:
!./max_pairwise_product_full< input1.txt > ctest1.txt
!grep -xvFf output1.txt ctest1.txt

90


In [35]:
%%writefile input2.txt
10
10 9 8 7 6 5 4 3 2 1

Overwriting input2.txt


In [36]:
%%writefile output2.txt
90

Overwriting output2.txt


In [37]:
!./max_pairwise_product_full< input2.txt > ctest2.txt
!grep -xvFf output2.txt ctest2.txt

### Test 2: largest input numbers 200000 200000 to check possible intger overflow

In [38]:
%%writefile input3.txt
6
1 3 200000 9 200000 23

Writing input3.txt


In [ ]:
200000*200000

40000000000

In [39]:
%%writefile output3.txt
40000000000

Writing output3.txt


In [40]:
!./max_pairwise_product_full < input3.txt

1345294336


In [41]:
!./max_pairwise_product_full < input3.txt > ctest3.txt
!grep -xvFf output3.txt ctest3.txt

1345294336


In [42]:
 %%writefile run_all_test.sh
#!/bin/bash
g++ max_pairwise_product_full.cpp -o max_pairwise_product_full
for i in {0..3}
do
    ./max_pairwise_product_full < input$i.txt > ctest$i.txt
    grep -xvFf output$i.txt ctest$i.txt
done

Overwriting run_all_test.sh


In [43]:
!chmod u+x run_all_test.sh
!./run_all_test.sh

90
1345294336


### Changing the variables to long long

In [44]:
%%writefile max_pairwise_product_long_full.cpp
#include <iostream>
#include <vector>
#include <algorithm>

long long MaxPairwiseProduct(const std::vector<int>& numbers) {
    long long max_product = 0LL;
	  long long aux_product = 0LL;
    int n = numbers.size();

    for (int first = 0; first < n; ++first) {
        for (int second = 0; second < n; ++second) {
          if ( first != second){
                aux_product = (long long) numbers[first] * numbers[second];
                if (aux_product > max_product)  max_product = aux_product;
          }
        }
    }


    return max_product;
}

int main() {
    int n;
    std::cin >> n;
    std::vector<int> numbers(n);
    for (int i = 0; i < n; ++i) {
        std::cin >> numbers[i];
    }

    std::cout << MaxPairwiseProduct(numbers) << "\n";
    return 0;
}


Writing max_pairwise_product_long_full.cpp


In [45]:
!g++ max_pairwise_product_long_full.cpp -o max_pairwise_product_long_full

In [ ]:
!./max_pairwise_product_long_full < input3.txt

40000000000


### Runing multuple test with a bash shell

In [46]:
 %%writefile run_all_test.sh
#!/bin/bash
g++ max_pairwise_product_long_full.cpp -o max_pairwise_product_long_full
for i in {1..3}
do
    ./max_pairwise_product_long_full < input$i.txt > ctest$i.txt
    grep -xvFf output$i.txt ctest$i.txt
done

Overwriting run_all_test.sh


In [47]:
!chmod u+x run_all_test.sh
!./run_all_test.sh

90


### Test 3  stress tests, 6 testing cases files input4.txt/output4.txt .. input10.txt/output10.txt

---



In [48]:
 %%writefile stress_test_generator.cpp
#include <iostream>
#include <vector>
#include <fstream>
#include <cstdlib> // for rand()
#include <ctime>   // for seeding srand()

using namespace std;

// Generate a random integer in a specified range
int random_int(int min, int max) {
    return min + rand() % (max - min + 1);
}

// Calculate the maximum pairwise product of a vector of integers
long long MaxPairwiseProduct(const vector<int>& numbers) {
    long long max_product = 0LL;
    int n = numbers.size();

    for (int first = 0; first < n; ++first) {
        for (int second = 0; second < n; ++second) {
            if (first != second) {
                long long aux_product = (long long)numbers[first] * numbers[second];
                if (aux_product > max_product) {
                    max_product = aux_product;
                }
            }
        }
    }

    return max_product;
}

// Generate a stress test
void generate_stress_test(int start_index, int num_tests, int min_size, int max_size, int min_value, int max_value) {
    srand(time(nullptr)); // Seed srand with current time

    for (int i = start_index; i < start_index + num_tests; ++i) {
        int n = random_int(min_size, max_size); // Generate random size for vector

        // Generate input file
        ofstream input_file("input" + to_string(i) + ".txt");
        input_file << n << "\n"; // Output vector size

        vector<int> a(n);
        for (int j = 0; j < n; ++j) {
            a[j] = random_int(min_value, max_value); // Generate random values for vector
            input_file << a[j] << " "; // Output vector element
        }
        input_file << "\n"; // End of vector
        input_file.close();

        // Generate output file
        ofstream output_file("output" + to_string(i) + ".txt");
        output_file << MaxPairwiseProduct(a) << "\n"; // Output result using MaxPairwiseProduct function
        output_file.close();
    }
}

int main() {
    int start_index, num_tests;

    // Request the start index and number of stress tests from the user
    std::cout << "Enter the start index for input files: ";
    std::cin >> start_index;
    std::cout << "Enter the number of stress tests to create: ";
    std::cin >> num_tests;

    // Generate stress tests starting from the specified start index
    generate_stress_test(start_index, num_tests, 2, 1000, 0, 10000);

    return 0;
}


Writing stress_test_generator.cpp


In [49]:
#stress tests randomly created
!g++ stress_test_generator.cpp -o stress_test_generator
!./stress_test_generator

#Start index: 4 for file4
#numbers of tests: 7 (from 4 to 10)

Enter the start index for input files: 4
Enter the number of stress tests to create: 7


In [ ]:
##input4.txt with expected output4.txt
!cat input4.txt
!cat output4.txt

481
3412 1542 1309 6805 8278 7232 6900 7611 90 874 7698 8631 5164 6281 1430 1251 1014 7251 2141 6750 519 4279 9638 8101 8452 8899 338 8775 6408 4150 661 9820 6771 3050 7703 6128 1360 4602 3738 1450 6555 1435 80 2797 8795 2589 4048 887 9840 7269 8716 359 1547 8353 9539 1078 8330 9877 9853 4737 5106 1593 4556 2955 5722 3337 9083 7082 7940 2820 8532 5573 5334 8612 8371 5207 2280 3497 6095 2119 765 4810 3557 3392 4242 3096 4470 2571 4051 5401 7309 9157 8073 2943 3191 3794 6281 2273 875 5299 6173 485 1950 2585 176 320 7793 2456 4897 4966 4575 6741 9776 8133 132 5096 2307 5681 7668 6358 1082 6055 6594 233 76 9785 4028 7436 3136 5982 2734 387 7547 4685 4052 7723 6084 1844 178 980 6810 5832 8801 7664 5043 8933 2760 7350 5693 1506 4787 6775 8640 1380 8087 8716 2243 3193 6152 6458 9176 9965 7925 6722 5728 1976 5523 1812 4899 5701 3871 2787 1532 2671 1529 7655 2683 5368 5004 8376 6874 869 6229 6592 2249 4315 5308 5571 8588 2538 3108 7763 2502 1032 5563 8231 4087 1085 1121 8986 7865 6071 1772 475 

In [ ]:
##input5.txt with expected output5.txt
!cat input5.txt
!cat output5.txt

199
6354 7638 8491 7210 8082 2379 8857 1864 7555 4593 1610 8637 7537 2993 8497 3752 9213 8254 7144 3530 2172 4117 8092 6966 4915 1169 8346 5681 3913 3830 3340 1346 2546 2909 9635 1707 5288 8491 3571 2843 4162 5181 2558 1698 9253 1054 6530 8465 386 3673 3074 3637 8869 2244 1681 3784 3414 26 543 8406 4936 4962 9752 8561 7871 9386 267 4238 8955 4917 8160 3116 1177 1796 5894 429 2850 3502 9974 3236 8254 3047 7953 8201 6370 9634 3063 9784 9661 3607 8190 5675 9648 9020 4235 7519 9485 5582 2835 8439 498 2073 2634 1675 3869 8528 3184 7798 3108 3157 2112 1361 7283 64 9562 3652 9699 2625 4515 9359 7311 3783 6112 6958 2802 1425 5555 2286 7007 8390 1804 8585 1541 4438 259 6489 4044 4522 4286 7152 7679 6399 8513 6040 6463 9153 771 6161 1777 5286 6598 9088 9069 2709 7125 1870 4135 2679 5236 1141 2148 7040 804 3689 1477 1064 178 5521 6665 5543 2672 5423 1941 2263 1462 9484 1415 2233 5644 4272 7519 2242 3359 7666 6030 483 615 1243 4242 5851 3464 6390 2890 4268 1157 5446 5332 2414 2045 
97585616


In [ ]:
##input6.txt with expected output6.txt
!cat input6.txt
!cat output6.txt

669
7958 5796 7420 977 8059 9961 460 552 3273 7184 4824 1870 9426 9263 9537 6534 824 151 7778 5066 6002 1241 2534 9971 5509 3692 6495 1920 6106 8540 4996 5142 4335 3494 6120 3472 3454 6580 5103 6727 3763 9928 8598 4267 268 9213 1880 1092 9364 9658 7238 6444 898 850 7493 7486 4542 3987 9406 1727 2526 5480 6869 7940 8974 2988 2490 2428 647 7593 233 4410 8599 9910 9757 8867 9122 1636 1038 9564 2372 9355 6007 4349 204 3499 1834 4747 7486 2319 6474 1090 7799 3342 108 6773 7410 2598 279 8057 1270 512 3545 9869 1501 3301 9815 622 6016 852 186 8388 206 7272 2736 1489 1850 4571 6236 414 6890 3788 1505 5767 8210 2692 2539 5619 6370 3897 4754 7640 4410 8299 7508 6990 2679 7322 8691 8695 9253 9956 7083 537 7228 897 2027 9078 5468 8263 9492 3436 3130 2075 9204 2418 4768 2821 9116 1137 6719 3869 9856 2207 3246 8442 9197 5925 6843 8966 4620 6095 8922 2781 6633 6149 3678 8660 6305 225 8001 6875 4740 1130 29 5022 3548 4797 7844 2663 7013 5641 7611 7947 7848 857 6388 8123 6782 3230 7088 2480 9326 6009 6

In [ ]:
##input7.txt with expected output7.txt
!cat input7.txt
!cat output7.txt

990
2959 6895 6640 1306 7595 6397 5156 3831 6255 8486 7224 1442 7352 4194 3810 1509 5179 1325 7803 4495 5324 1850 3937 7318 4835 4638 6219 1677 2598 2758 4564 5557 732 1203 6863 8327 8679 3098 2157 6012 1583 460 7454 13 4654 2342 2601 911 4746 1483 6485 1149 4412 1500 8467 9247 6139 5764 2002 8737 9602 6566 5372 333 7769 2235 9739 7526 6412 1896 3537 9074 3435 2069 9087 9168 4411 2766 78 9158 4249 7642 1385 8661 9142 9852 8986 6359 5616 987 6174 6296 8632 1546 6629 6400 4860 6367 3925 1271 9342 8541 1423 2776 609 1588 3022 5021 4354 3100 5257 9683 1820 6642 9422 2041 7572 8408 8400 4266 473 4574 561 9106 7199 8269 6584 2058 5715 1588 4408 5056 128 5831 8912 738 7419 1933 6838 1772 6113 3173 2533 9012 893 1955 1052 8465 362 9453 3810 835 5105 4371 9941 2303 3719 6525 5440 9434 8113 9848 5568 8241 5678 4479 57 4175 7492 6895 5947 4683 67 9560 3694 960 1514 5826 504 1876 6357 4314 3790 2540 9764 3731 5922 3482 1334 1361 2915 525 2287 9563 9845 7965 5120 9903 3218 2611 7876 243 8373 7944 9

In [ ]:
##input8.txt with expected output8.txt
!cat input8.txt
!cat output8.txt

893
6756 6432 5022 4991 6521 9112 5366 161 6407 2197 7595 3150 4289 9176 7387 6012 2161 453 201 6572 1639 4561 3380 8024 8407 7207 3147 8986 2054 9939 7332 8810 6371 2353 3800 3970 2544 244 4131 8951 3521 2804 3180 8889 1979 566 4900 5219 1019 5101 1791 3738 9663 5171 1761 9148 2377 4908 9213 4431 4847 6544 3240 1217 8897 8120 6266 2519 8364 397 1470 1884 4280 5729 772 6260 6295 6752 1478 8393 2931 4348 2130 2593 9520 3891 2820 1896 9879 3111 6328 4725 733 646 7021 9630 8766 3286 3228 7130 4762 4698 92 9043 426 1944 6381 7800 8696 8938 6192 1626 3286 9402 5299 3884 4371 9198 5780 4249 3387 3186 52 4120 4912 8152 4828 3677 2517 8056 1885 7279 2753 3057 7400 4258 5001 4859 3136 4775 3797 9329 7480 8162 9809 3857 3124 4179 4133 8904 9508 7520 3169 638 1639 9160 8791 6468 3915 1307 4523 5801 9665 8356 9937 7065 3692 4937 1923 6829 790 6799 7236 8270 6039 7044 3206 9163 2301 7339 9146 1808 4859 3393 2447 6498 2552 2316 4044 6467 3623 8568 3346 3287 8002 3282 1430 1693 9298 4433 8522 87 1231

In [ ]:
##input9.txt with expected output9.txt
!cat input9.txt
!cat output9.txt

420
2505 2940 1961 409 4868 3511 490 4504 6923 7866 7678 609 5008 8138 8826 7268 9886 8405 8552 7731 4392 7518 8013 9121 788 9496 1293 1807 22 2812 8352 3606 6831 312 5095 2777 4902 5585 7281 1825 4530 6038 3513 616 5254 2339 7884 5139 1822 7514 2869 6214 5031 881 5335 6898 1455 7707 9784 2557 1597 8135 7242 9507 9526 2336 2284 5507 9001 643 7332 4609 7760 844 5225 3013 4262 4188 9231 6084 1701 2099 2298 7812 4059 8712 4709 6594 7497 5572 229 9095 4785 7471 8601 5390 886 1963 896 965 2607 9307 5574 366 150 1877 4459 5492 6065 3689 1575 7767 5789 4952 5578 926 4742 1365 7520 2239 8016 8828 2412 3880 6299 1012 9270 8264 4055 165 9229 6662 9472 5881 8107 700 7758 2565 6192 3823 6255 8847 2668 3122 3798 9325 4048 8541 689 2647 1858 9785 1474 4270 3664 7773 6361 2933 7115 415 4177 6343 7077 4727 2223 6263 5427 9982 8828 2698 4883 5082 1544 7551 8204 6421 6875 3331 6040 8643 5978 7898 9506 7452 3246 3169 6304 9608 7181 3418 22 1357 840 8179 6084 4142 4441 2590 5202 3268 6367 84 9430 8990 871

In [ ]:
##input10.txt with expected output5]10.txt
!cat input10.txt
!cat output10.txt

829
3471 5343 164 2903 8935 1197 1040 7323 9917 6434 2572 1945 4689 6952 2757 6652 5926 2594 9088 3977 3317 4803 6176 1258 3060 588 4169 2665 2088 6787 412 5560 2129 576 9542 2142 2852 582 9466 2768 7016 3116 5792 2783 67 9629 513 5993 2222 9602 1049 6618 4404 7225 7876 7464 8892 3123 1208 979 9910 1620 7618 2039 3276 7160 5260 7207 7742 4725 1054 5836 8920 6846 9698 65 6474 211 7138 9775 891 8187 7471 5295 6490 5346 3837 5381 8470 5045 7439 9458 7745 6136 2575 1020 4374 7836 9306 3194 3639 359 108 2559 8285 9806 2624 4758 1095 9762 5612 1986 9027 3082 8360 5516 9508 2197 896 7977 8321 9415 8513 6065 5550 1088 8164 1002 8924 7470 4196 3641 8908 4304 6200 7192 4109 9904 1950 6284 743 8641 8270 9771 1722 7709 5286 1229 9906 7262 284 8226 7755 8798 5370 4383 964 3533 5385 966 2081 9581 4607 989 3884 1886 8181 9072 1789 1209 6434 2532 9850 5783 3381 2651 3491 9747 4959 3396 7008 6323 2700 4762 6199 8070 223 7163 1603 5608 9208 3684 5188 3814 5752 150 5700 3933 300 8568 5142 6735 2179 6071 

### Test 4  time test  measure the execution time running with the the largest worst cases input11.txt  (200000 increasing numbers) and input12.txt (199998 increasing numbers and  200000, 200000 at the end)

In [50]:
# input11.txt 200000 increasing numbers:  1 2 .. 200000

%%writefile WriteTest.cpp

#include <iostream>
#include <fstream>
#include <cstdlib>

using namespace std;

int main() {
  // Create and open a text file
  ofstream MyFile("input11.txt");

 MyFile << 200000 <<"\n";  // n
 for (int i=1; i <200001; i++){
     MyFile << i <<" ";
 }

  // Close the file
  MyFile.close();
}

Writing WriteTest.cpp


In [51]:
!g++ WriteTest.cpp -o WriteTest
!./WriteTest

In [ ]:
199999 * 200000

39999800000

In [52]:
%%writefile output11.txt
39999800000

Writing output11.txt


In [53]:
# input12.txt 200000 decreasing numbers and 200000, 199999 ...  2 1

%%writefile WriteTest.cpp

#include <iostream>
#include <fstream>
#include <cstdlib>

using namespace std;

int main() {
  // Create and open a text file
  ofstream MyFile("input12.txt");

 MyFile << 200000 <<"\n";  // n
 for (int i=1; i <200001; i++){
     MyFile << 200001-i <<" ";
 }

  // Close the file
  MyFile.close();
}

Overwriting WriteTest.cpp


In [54]:
!g++ WriteTest.cpp -o WriteTest
!./WriteTest

In [55]:
%%writefile output12.txt
39999800000

Writing output12.txt


### Measure time

https://askubuntu.com/questions/53444/how-can-i-measure-the-execution-time-of-a-terminal-process

In [56]:
!g++ max_pairwise_product_long_full.cpp -o max_pairwise_product_long_full
!time ./max_pairwise_product_long_full< input11.txt

39999800000

real	5m8.502s
user	5m3.650s
sys	0m0.144s


In [ ]:
!g++ max_pairwise_product_long_full.cpp -o max_pairwise_product_long_full
!time ./max_pairwise_product_long_full< input12.txt

39999800000

real	4m46.585s
user	4m43.714s
sys	0m0.110s


### Calculate the number or and upper bound to the number of instructions executed by the algorithm as a function of n, add in each line a comment with the number of instructions that are executed in that line.

In [57]:
%%writefile max_pairwise_product_long_full.cpp
#include <iostream>
#include <vector>
#include <algorithm>

long long MaxPairwiseProduct(const std::vector<int>& numbers) {              //INSTRUCTIONS executed by line
    long long max_product = 0LL;                                             //1
	  long long aux_product = 0LL;                                             //1
    int n = numbers.size();                                                  //1

    for (int first = 0; first < n; ++first) {                                //2+2n
        for (int second = 0; second < n; ++second) {                         //2n + 2n**2
          if ( first != second ){                                            //n**2
                aux_product = numbers[first] * numbers[second];              //n**2 -n
                if (aux_product > max_product)                               //n**2 -n
                    max_product = aux_product;                               //[0, n**2/2 - n/2] upper bound
          }
        }
    }

                                                                             // --------------
    return max_product;                                                      // 5 + 2*n + [5*n**2, 11*n**2/2  - n/2]  upper bound
}

int main() {                                              //INSTRUCTIONS executed by line
    int n;                                                //1
    std::cin >> n;                                        //1
    std::vector<int> numbers(n);                          //n
    for (int i = 0; i < n; ++i) {                         //2 + 2n
        std::cin >> numbers[i];                           //n
    }

    std::cout << MaxPairwiseProduct(numbers) << "\n";     //1 + 1  + 5 + 2*n + [5*n**2, 11*n**2/2  - n/2]

    return 0;                                             // --------------
}                                                         //  11 + 6*n + [5*n**2, 11*n**2/2  - n/2]  upper bound


Overwriting max_pairwise_product_long_full.cpp


### Count executed instructions in code

In [ ]:
%%writefile max_pairwise_product_long_full_execints.cpp
#include <iostream>
#include <vector>
#include <algorithm>

long long MaxPairwiseProduct(const std::vector<int>& numbers, long long &execints) {      //INSTRUCTIONS executed by line

    long long max_product = 0LL;                                                          //1
    execints++;

	  long long aux_product = 0LL;                                                          //1
    execints++;

    int n = numbers.size();                                                               //1
    execints++;

    execints++;
    for (int first = 0; first < n; ++first) {                                             //2+2n
        execints += 2;

        execints++;
        for (int second = 0; second < n; ++second) {                                       //2n + 2n**2
          execints += 2;

          execints++;
          if ( first != second ){                                                         //n**2
                aux_product = numbers[first] * numbers[second];                           //n**2 -n
                execints++;

                execints++;
                if (aux_product > max_product){                                           //n**2 -n
                    max_product = aux_product;                                            //[0,n**2/2 - n/2)  upper bound
                    execints++;

                }
        }

      }
      execints++;
    }
    execints++;
                                                                                        // --------------
    return max_product;                                                                 // 5 + 2*n + [5*n**2, 11*n**2/2  - n/2] upper bound
}

int main() {                                                 //INSTRUCTIONS executed by line

    long long execints =0LL;

    int n;                                                   //1
    execints++;

    std::cin >> n;                                           //1
    execints++;

    std::vector<int> numbers(n);                             //n
    execints += n;

    execints++;
    for (int i = 0; i < n; ++i) {                            //2 + 2n
        execints += 2;

        std::cin >> numbers[i];
        execints++;                                          //n

    }


    //std::cout << MaxPairwiseProduct(numbers) << "\n";      //1 + 1  + 5 + 2*n + [5*n**2, 11*n**2/2  - n/2] upper bound

    execints++;
    int M = MaxPairwiseProduct(numbers,execints);

    execints++;
    std::cout <<  execints << "\n";
    return 0;                                                // --------------
}                                                            // 11 + 6*n + [5*n**2, 11*n**2/2  - n/2]


Overwriting max_pairwise_product_long_full_execints.cpp


The number of instructions executed by the algorithm with a list of size $n$ is  

$$T (n) = 11 + 6n + \left[5n^2, 11\frac{n^2}{2}  - \frac{n}{2}\right)  >  6n^2 = O(n^2)$$

It depends on the size of the lists and  the order of the list.

For the input0.txt
```
10
0 0 0 0 0 0 0 0 0 0
```
$$T = 11 + 6(10) + 5(10)^2   = 561 $$

$$T(10) \in [571, 621)  $$


In [ ]:
!g++ max_pairwise_product_long_full_execints.cpp -o max_pairwise_product_long_full_execints
!./max_pairwise_product_long_full_execints < input0.txt

570


For the input1.txt
```
10
1 2 3 4 5 6 7 8 9 10
```
$$T = 11 + 6(10) + 5(10)^2 +  9 + 5 + 4 + 3 + 2 + 2 + 2 + 2 + 1 + 0  = 601 $$

$$T(10) \in [572, 622)  $$


In [ ]:
#The number of times that  max_product = aux_product; in the worst case is less than (n**2 -n)/2 -  upper bound

#     1  2  3  4  5  6  7  8  9  10
#  1  x  2  3  4  5  6  7  8  9  10         9
#  2  2  x  6  8  10 12 14 16 18 20         5 14
#  3  3  6  x  12 15 18 21 24 27 30         4 18
#  4  4  8  12 x  20 24 28 32 36 40         3 21
#  5  5  10 15 20 x  30 35 40 45 50         2 23
#  6  6  12 18 24 30 x  42 48 54 60         2 25
#  7  7  14 21 28 35 42 x  56 63 70         2 27
#  8  8  18 27 36 45 54 63 72 x  90         1 30
#  10 10 20 30 40 50 60 70 60 90 x          0

In [ ]:
11 + 6*(10) + 5*(10)*(10) +  9 + 5 + 4 + 3 + 2 + 2 + 2 + 2 + 1 + 0

601

In [ ]:
n = 10
print( "[", 11 + 6*n + 5*n**2,  ",", 11 + 6*n  + 5.5*n**2, ")")

[ 571 , 621.0 )


In [ ]:
!g++ max_pairwise_product_long_full_execints.cpp -o max_pairwise_product_long_full_execints
!./max_pairwise_product_long_full_execints < input1.txt

/bin/bash: line 1: input1.txt: No such file or directory



For the input2.txt
```
10
10 9 8 7 6 5 4 3 2 1
```
$$T = 11 + 6*(10) + 5*(10)**2 +  1  = 571 $$

In [ ]:
11 + 6*(10) + 5*(10)**2 +  1

572

In [ ]:
!g++ max_pairwise_product_long_full_execints.cpp -o max_pairwise_product_long_full_execints
!./max_pairwise_product_long_full_execints < input2.txt

571


Calculate estimated time by instruction and number of instructions by second

For the input11.txt
```
200000
1 2 ..  199999 2000
```
$$T = 11 + 6*(200000) + 5.5*(200000)**2  = 200001200012  $$

In [ ]:
!g++ max_pairwise_product_long_full_execints.cpp -o max_pairwise_product_long_full_execints
!./max_pairwise_product_long_full_execints < input11.txt

^C


In [ ]:
# Estimated time by instruction

295/200003177091


1.474976569326082e-09

In [ ]:
# Estimated number of instructions by second

200003177091/295

677976871.4949152

### Estimate and verify the maximun n that can be solved in 1 second

$$(5.5n^2 + 5.5n + 11) \times 1.47*10^{-9} s = 1 s$$

$$ n \approx 11000$$


In [ ]:
# Verify input13.txt
%%writefile WriteTest.cpp

#include <iostream>
#include <fstream>
#include <cstdlib>

using namespace std;

int main() {
  // Create and open a text file
  ofstream MyFile("input13.txt");

 MyFile << 11000 <<"\n";  // n
 for (int i=1; i <11001; i++){
     MyFile << i <<" ";
 }

  // Close the file
  MyFile.close();
}

Overwriting WriteTest.cpp


In [ ]:
!g++ WriteTest.cpp -o WriteTest
!./WriteTest

In [ ]:
10999 * 11000

120989000

In [ ]:
!g++ max_pairwise_product_long_full.cpp -o max_pairwise_product_long_full
!time ./max_pairwise_product_long_full< input13.txt

120989000

real	0m1.411s
user	0m1.360s
sys	0m0.004s


## Python

In [ ]:
# Verify

# For the following solutions required  the same tests and analysis performed for the first solution

## Solutions in C++ and Python that uses
$$\frac{n^2 - n}{2} = O(n^2)\ \ \text{products}$$

In [ ]:
%%writefile max_pairwise_product.cpp
#include <iostream>
#include <vector>
#include <algorithm>

int MaxPairwiseProduct(const std::vector<int>& numbers) {
    int max_product = 0;
    int n = numbers.size();

    for (int first = 0; first < n; ++first) {
        for (int second = first + 1; second < n; ++second) {
            max_product = std::max(max_product,
                numbers[first] * numbers[second]);
        }
    }

    return max_product;
}

int main() {
    int n;
    std::cin >> n;
    std::vector<int> numbers(n);
    for (int i = 0; i < n; ++i) {
        std::cin >> numbers[i];
    }

    std::cout << MaxPairwiseProduct(numbers) << "\n";
    return 0;
}



Overwriting max_pairwise_product.cpp


In [ ]:
%%writefile max_pairwise_product_long.cpp
#include <iostream>
#include <vector>
#include <algorithm>

long long MaxPairwiseProduct(const std::vector<int>& numbers) {              //INSTRUCTIONS executed by line
    long long max_product = 0LL;                                             //1
	  long long aux_product = 0LL;                                             //1
    int n = numbers.size();                                                  //1

    for (int first = 0; first < n; ++first) {                                //2+2n
        for (int second = first + 1; second < n; ++second) {                 //2n + (n-1)n
			       aux_product =  (long long) numbers[first] * numbers[second];    //(n-1)n/2
            if (max_product < aux_product)  {                                //(n-1)n/2
				        max_product = aux_product;                                   //[1,(n-1)n/2)  upper bound
             }
        }
    }
                                                                             // --------------
    return max_product;                                                      // 5 + 4n + [2(n-1)n+1, 2.5 (n-1)n) upper bound
}

int main() {                                              //INSTRUCTIONS executed by line
    int n;                                                //1
    std::cin >> n;                                        //1
    std::vector<int> numbers(n);                          //n
    for (int i = 0; i < n; ++i) {                         //2 + 2n
        std::cin >> numbers[i];                           //n
    }

    std::cout << MaxPairwiseProduct(numbers) << "\n";     //1 + 1  + 5 + 4n + [2(n-1)n+1, 2.5 (n-1)n) upper bound
    return 0;                                             // --------------
}                                                         // 11 + 7n + [2(n-1)n+1, 2.5 (n-1)n) upper bound


Overwriting max_pairwise_product_long.cpp


### Estimate and verify the maximum n that can be solved in 1 second



In [ ]:
%%writefile max_n_in_a_second.cpp
#include <iostream>
#include <vector>
#include <chrono>

// Function to calculate the maximum pairwise product
long long MaxPairwiseProduct(const std::vector<int>& numbers) {
    long long max_product = 0LL;
    int n = numbers.size();

    for (int first = 0; first < n; ++first) {
        for (int second = first + 1; second < n; ++second) {
            long long aux_product = (long long)numbers[first] * numbers[second];
            if (aux_product > max_product) {
                max_product = aux_product;
            }
        }
    }

    return max_product;
}

int main() {
    int n = 10; // Initial value of n
    auto start_time = std::chrono::high_resolution_clock::now();

    // Loop to find the maximum n that can be solved in 1 second
    while (true) {
        // Generate input data of size n
        std::vector<int> input_data(n, 1); // Initialize input data with all elements as 1

        // Measure the execution time
        auto start = std::chrono::high_resolution_clock::now();
        MaxPairwiseProduct(input_data);
        auto end = std::chrono::high_resolution_clock::now();

        // Calculate the elapsed time
        double elapsed_time = std::chrono::duration_cast<std::chrono::milliseconds>(end - start).count() / 1000.0;

        // Check if the elapsed time exceeds 1 second
        if (elapsed_time >= 1.0) {
            break; // Stop the loop if elapsed time exceeds 1 second
        }

        // Double the value of n for the next iteration
        n *= 2;
    }

    auto end_time = std::chrono::high_resolution_clock::now();
    double total_time = std::chrono::duration_cast<std::chrono::milliseconds>(end_time - start_time).count() / 1000.0;

    // Print the maximum n that can be solved in 1 second
    std::cout << "Maximum n that can be solved in 1 second: " << n << std::endl;
    std::cout << "Total execution time: " << total_time << " seconds" << std::endl;

    return 0;
}


Overwriting max_n_in_a_second.cpp


In [ ]:
!g++ -o max_n_in_a_second max_n_in_a_second.cpp
!./max_n_in_a_second

Maximum n that can be solved in 1 second: 20480
Total execution time: 1.829 seconds


In [80]:
%%writefile  stress_test_of_size_n.cpp

#include <iostream>
#include <vector>
#include <fstream>
#include <cstdlib> // for rand()
#include <ctime>   // for seeding srand()

using namespace std;

// Generate a random integer in a specified range
int random_int(int min, int max) {
    return min + rand() % (max - min + 1);
}

// Generate a stress test of size n
void generate_stress_test(int n, int min_value, int max_value) {
    srand(time(nullptr)); // Seed srand with current time

    // Generate input file
    ofstream input_file("stress_test_of_size_n.txt");
    input_file << n << "\n"; // Output vector size

    // Generate random values for vector
    for (int j = 0; j < n; ++j) {
        int random_value = random_int(min_value, max_value);
        input_file << random_value << " "; // Output vector element
    }

    input_file << "\n"; // End of vector
    input_file.close();
}

int main() {
    int n;

    // Request the size of the vector (n) from the user
    cout << "Enter the size of the vector (n): ";
    cin >> n;

    int min_value = 0;     // Minimum value for random integers
    int max_value = 200000; // Maximum value for random integers

    // Generate a single stress test with specified size
    generate_stress_test(n, min_value, max_value);

    return 0;
}


Writing stress_test_of_size_n.cpp


In [ ]:
#Create stress test of size n = 20480
!g++ -o stress_test_of_size_n stress_test_of_size_n.cpp
!./stress_test_of_size_n

Enter the size of the vector (n): 20480


In [ ]:
#Verify that max_wisepair_product_long can run an input of size n = 20480 in a second
!g++ -o max_pairwise_product_long max_pairwise_product_long.cpp
!time ./max_pairwise_product_long < stress_test_of_size_n.txt

39994800133

real	0m2.116s
user	0m2.057s
sys	0m0.006s


# Sort solution in C++ and Python

$$ O(n \lg n)$$

**Which sorting algorithm is used in GCC?**
https://stackoverflow.com/questions/7221257/which-sorting-algorithm-is-used-in-gcc

* ...  the algorithm is called introsort, a combination between Insertionsort, Quicksort & Heapsort. ...

**Introsort**
https://en.wikipedia.org/wiki/Introsort

Introsort or introspective sort is a hybrid sorting algorithm that provides both fast average performance and (asymptotically) optimal worst-case performance. It begins with quicksort, it switches to heapsort when the recursion depth exceeds a level based on (the logarithm of) the number of elements being sorted and it switches to insertion sort when the number of elements is below some threshold

*   Class	Sorting algorithm
*   Data structure	Array
*   Worst-case performance	$O(n \lg n)$
*   Average performance	$O(n \lg n)$

**Timsort**
https://en.wikipedia.org/wiki/Timsort

Timsort is a hybrid, stable sorting algorithm, derived from merge sort and insertion sort, designed to perform well on many kinds of real-world data. It was implemented by Tim Peters in 2002 for use in the Python programming language.

*   Class	Sorting algorithm
*   Data structure	Array
*   Worst-case performance	$O(n \lg n)$
*   Best-case performance $O(n)$
*   Average performance	$O(n \lg n)$
*   Worst-case space complexity $O(n)$

In [58]:
%%writefile max_pairwise_product_long_sorting.cpp

#include <iostream>
#include <vector>
#include <algorithm>
#include <cmath> // Include cmath for std::log

// Implement introsort
template<typename RandomIt>
void introsort(RandomIt first, RandomIt last, int depth_limit) {
    while (last - first > 16) {
        if (depth_limit == 0) {
            std::make_heap(first, last);
            std::sort_heap(first, last);
            return;
        }
        --depth_limit;
        RandomIt pivot = std::partition(first, last, [=](const auto& em){return em < *(last - 1);});
        std::swap(*pivot, *(last - 1));
        introsort(first, pivot, depth_limit);
        first = pivot + 1;
    }
    std::sort(first, last);
}

// Wrapper function to use introsort
template<typename RandomIt>
void introsort(RandomIt first, RandomIt last) {
    introsort(first, last, 2 * std::log(last - first));
}

long long MaxPairwiseProduct(const std::vector<int>& numbers) {
    std::vector<int> sorted_numbers = numbers;
    introsort(sorted_numbers.begin(), sorted_numbers.end());

    int n = sorted_numbers.size();
    return (long long)sorted_numbers[n - 1] * sorted_numbers[n - 2];
}

int main() {
    int n;
    std::cin >> n;
    std::vector<int> numbers(n);
    for (int i = 0; i < n; ++i) {
        std::cin >> numbers[i];
    }

    std::cout << MaxPairwiseProduct(numbers) << "\n";
    return 0;
}



Writing max_pairwise_product_long_sorting.cpp


Test with inputs 1 and 2



In [59]:
!g++ -o max_pairwise_product_long_sorting max_pairwise_product_long_sorting.cpp
!./max_pairwise_product_long_sorting < input1.txt
!./max_pairwise_product_long_sorting < input2.txt


90
90


Test and time it with input 3

In [60]:
!g++ -o max_pairwise_product_long_sorting max_pairwise_product_long_sorting.cpp
!time ./max_pairwise_product_long_sorting < input3.txt


40000000000

real	0m0.004s
user	0m0.000s
sys	0m0.003s


In [61]:
#Test with input 11
!g++ -o max_pairwise_product_long max_pairwise_product_long.cpp
!time ./max_pairwise_product_long < input11.txt

cc1plus: fatal error: max_pairwise_product_long.cpp: No such file or directory
compilation terminated.
/bin/bash: line 1: ./max_pairwise_product_long: No such file or directory

real	0m0.001s
user	0m0.001s
sys	0m0.000s


20 stress tests

In [62]:
#stress tests randomly created with index from 13 to 32 for files
!g++ stress_test_generator.cpp -o stress_test_generator
!./stress_test_generator

#Start index = 13
# Number of tests: 20

Enter the start index for input files: 13
Enter the number of stress tests to create: 20


In [63]:
%%writefile run_all_test.sh
#!/bin/bash
g++ max_pairwise_product_full.cpp -o max_pairwise_product_full
for i in {13..32}
do
    ./max_pairwise_product_full < input$i.txt > ctest$i.txt
    echo "Testing on input$i"
    time ./max_pairwise_product_full < input$i.txt
    echo ""
    grep -xvFf output$i.txt ctest$i.txt
done

Overwriting run_all_test.sh


In [64]:
!chmod u+x run_all_test.sh
!./run_all_test.sh

Testing on input13
99560340

real	0m0.008s
user	0m0.005s
sys	0m0.001s

Testing on input14
99550500

real	0m0.006s
user	0m0.005s
sys	0m0.001s

Testing on input15
98922835

real	0m0.003s
user	0m0.002s
sys	0m0.001s

Testing on input16
99510220

real	0m0.003s
user	0m0.002s
sys	0m0.001s

Testing on input17
99770126

real	0m0.005s
user	0m0.003s
sys	0m0.002s

Testing on input18
99780040

real	0m0.004s
user	0m0.004s
sys	0m0.000s

Testing on input19
95411457

real	0m0.002s
user	0m0.002s
sys	0m0.000s

Testing on input20
98316270

real	0m0.003s
user	0m0.003s
sys	0m0.000s

Testing on input21
99340765

real	0m0.005s
user	0m0.004s
sys	0m0.002s

Testing on input22
99311190

real	0m0.010s
user	0m0.008s
sys	0m0.001s

Testing on input23
99660264

real	0m0.005s
user	0m0.004s
sys	0m0.001s

Testing on input24
99840039

real	0m0.011s
user	0m0.008s
sys	0m0.002s

Testing on input25
99231176

real	0m0.004s
user	0m0.003s
sys	0m0.001s

Testing on input26
97277928

real	0m0.002s
user	0m0.001s
sys	0m0.001s

Testin

Count executed instructions in code and verify with inputs 1,2 and 3


In [65]:
%%writefile max_pairwise_product_long_sorting_execints.cpp

#include <iostream>
#include <vector>
#include <algorithm>
#include <cmath> // Include cmath for std::log

long long execints = 0LL;

// Implement introsort
template<typename RandomIt>
void introsort(RandomIt first, RandomIt last, int depth_limit) {
    while (last - first > 16) {
        if (depth_limit == 0) {
            std::make_heap(first, last);
            std::sort_heap(first, last);
            return;
        }
        --depth_limit;
        RandomIt pivot = std::partition(first, last, [=](const auto& em){return em < *(last - 1);});
        std::swap(*pivot, *(last - 1));
        introsort(first, pivot, depth_limit);
        first = pivot + 1;
    }
    std::sort(first, last);
}

// Wrapper function to use introsort
template<typename RandomIt>
void introsort(RandomIt first, RandomIt last) {
    introsort(first, last, 2 * std::log(last - first));
}

long long MaxPairwiseProduct(const std::vector<int>& numbers) {
    std::vector<int> sorted_numbers = numbers;
    introsort(sorted_numbers.begin(), sorted_numbers.end());

    int n = sorted_numbers.size();
    execints += 2; // For assignment and size check
    execints += 2 * std::log(n); // For sorting

    return (long long)sorted_numbers[n - 1] * sorted_numbers[n - 2];
}

int main() {
    int n;
    std::cin >> n;
    execints += 1; // For reading n
    std::vector<int> numbers(n);
    for (int i = 0; i < n; ++i) {
        std::cin >> numbers[i];
        execints += 2; // For reading and incrementing i
    }

    //std::cout << MaxPairwiseProduct(numbers) << "\n";
    std::cout << execints << "\n"; // Output the count of executed instructions
    return 0;
}


Writing max_pairwise_product_long_sorting_execints.cpp


In [72]:
#Compute the time using some of the input files
!g++ max_pairwise_product_long_sorting_execints.cpp -o max_pairwise_product_long_sorting_execints
!time ./max_pairwise_product_long_sorting_execints < input11.txt  > output11.txt
!time ./max_pairwise_product_long_sorting_execints < input25.txt  > output25.txt
!time ./max_pairwise_product_long_sorting_execints < input32.txt  > output32.txt



real	0m0.073s
user	0m0.066s
sys	0m0.005s

real	0m0.005s
user	0m0.000s
sys	0m0.004s

real	0m0.004s
user	0m0.001s
sys	0m0.002s


Calculate estimated time by instruction and number of instructions by second

In [75]:
# Read the content of output files
with open('output11.txt', 'r') as file:
    output11_content = file.read().strip()

with open('output25.txt', 'r') as file:
    output2_content = file.read().strip()

with open('output32.txt', 'r') as file:
    output3_content = file.read().strip()

# Parse the number of executed instructions
executed_instructions_1 = int(output1_content)
executed_instructions_2 = int(output2_content)
executed_instructions_3 = int(output3_content)

# Real-time measurements in seconds
real_time_1 = 0.073
real_time_2 = 0.003
real_time_3 = 0.004

# Compute Estimated Time per Instruction (ETI) and Instructions per Second (IPS)
eti_1 = real_time_1 / executed_instructions_1
ips_1 = executed_instructions_1 / real_time_1

eti_2 = real_time_2 / executed_instructions_2
ips_2 = executed_instructions_2 / real_time_2

eti_3 = real_time_3 / executed_instructions_3
ips_3 = executed_instructions_3 / real_time_3

# Print the results
print("Output 1:")
print("ETI: {:.9f} s/instruction".format(eti_1))
print("IPS: {:.0f} instructions/second".format(ips_1))

print("\nOutput 2:")
print("ETI: {:.9f} s/instruction".format(eti_2))
print("IPS: {:.0f} instructions/second".format(ips_2))

print("\nOutput 3:")
print("ETI: {:.9f} s/instruction".format(eti_3))
print("IPS: {:.0f} instructions/second".format(ips_3))



Output 1:
ETI: 0.000000182 s/instruction
IPS: 5479466 instructions/second

Output 2:
ETI: 0.000003161 s/instruction
IPS: 316333 instructions/second

Output 3:
ETI: 0.000003033 s/instruction
IPS: 329750 instructions/second


### Estimate and verify the maximum n that can be solved in 1 second

In [76]:
%%writefile max_n_in_a_second_sorting.cpp

#include <iostream>
#include <vector>
#include <algorithm>
#include <cmath> // Include cmath for std::log
#include <chrono> // Include chrono for timing

// Implement introsort
template<typename RandomIt>
void introsort(RandomIt first, RandomIt last, int depth_limit) {
    while (last - first > 16) {
        if (depth_limit == 0) {
            std::make_heap(first, last);
            std::sort_heap(first, last);
            return;
        }
        --depth_limit;
        RandomIt pivot = std::partition(first, last, [=](const auto& em){return em < *(last - 1);});
        std::swap(*pivot, *(last - 1));
        introsort(first, pivot, depth_limit);
        first = pivot + 1;
    }
    std::sort(first, last);
}

// Wrapper function to use introsort
template<typename RandomIt>
void introsort(RandomIt first, RandomIt last) {
    introsort(first, last, 2 * std::log(last - first));
}

long long MaxPairwiseProduct(const std::vector<int>& numbers) {
    std::vector<int> sorted_numbers = numbers;
    introsort(sorted_numbers.begin(), sorted_numbers.end());

    int n = sorted_numbers.size();
    return (long long)sorted_numbers[n - 1] * sorted_numbers[n - 2];
}

// Function to measure execution time for a given input size
double measureTimeForN(int n) {
    std::vector<int> numbers(n, 1); // Generate input with all elements as 1
    auto start = std::chrono::high_resolution_clock::now(); // Start the timer
    MaxPairwiseProduct(numbers); // Call the function
    auto end = std::chrono::high_resolution_clock::now(); // Stop the timer
    std::chrono::duration<double> elapsed_seconds = end - start; // Calculate duration
    return elapsed_seconds.count(); // Return the duration in seconds
}

// Function to find the maximum n that can be solved in 1 second
int findMaxNWithinTimeLimit(double time_limit) {
    int low = 100; // Initial low value of n
    int high = low; // Initial high value of n

    // Increase high until execution time exceeds 1 second
    while (measureTimeForN(high) <= time_limit) {
        low = high;
        high *= 2;
    }

    // Binary search to find the maximum n within the time limit
    while (low <= high) {
        int mid = low + (high - low) / 2;
        if (measureTimeForN(mid) <= time_limit) {
            low = mid + 1; // Move to the right half
        } else {
            high = mid - 1; // Move to the left half
        }
    }

    return high; // Return the last successful value of n
}

int main() {
    double time_limit = 1.0; // Time limit in seconds
    int max_n = findMaxNWithinTimeLimit(time_limit);
    std::cout << "Maximum n that can be solved in 1 second: " << max_n << std::endl;
    return 0;
}


Writing max_n_in_a_second_sorting.cpp


In [77]:
!g++ -o max_n_in_a_second_sorting max_n_in_a_second_sorting.cpp
!./max_n_in_a_second_sorting


Maximum n that can be solved in 1 second: 551997


In [81]:
#Create stress test of size n = 551997
!g++ -o stress_test_of_size_n stress_test_of_size_n.cpp
!./stress_test_of_size_n

Enter the size of the vector (n): 551997


In [82]:
#Verify the max_pairwise_product_long_sorting can run stress test of size n = 551997

!g++ -omax_pairwise_product_long_sorting max_pairwise_product_long_sorting.cpp
!time ./max_pairwise_product_long_sorting < stress_test_of_size_n.txt

40000000000

real	0m0.649s
user	0m0.632s
sys	0m0.010s


## Efficient solution with largest and second-largest in an list, array or vector in C++ and Python

 $$T(n) < 10 n  = O(n)$$





Add in each line a comment with the number of instructions that are executed in that line

In [86]:
%%writefile max_pairwise_product_long_efficient_execints.cpp
#include <iostream>
#include <vector>

std::pair<long long, long long> MaxPairwiseProduct(const std::vector<int>& numbers, long long& execints) { // 1 instruction
    long long max_product = 0; // 1 instruction
    execints += 1; // increment instruction count

    long long second_max = 0; // 1 instruction
    execints += 1; // increment instruction count

    for (int i = 0; i < numbers.size(); ++i) { // 3 instructions + (n + 1) * 2 instructions
        execints += 3; // increment instruction count
        execints += 2; // increment instruction count for comparison
        if (numbers[i] > max_product) { // 2 instructions
            second_max = max_product; // 1 instruction
            execints += 1; // increment instruction count
            max_product = numbers[i]; // 1 instruction
            execints += 1; // increment instruction count
        } else if (numbers[i] > second_max) { // 2 instructions
            second_max = numbers[i]; // 1 instruction
            execints += 1; // increment instruction count
        }
    }

    return std::make_pair(max_product, second_max); // 1 instruction
}

int main() { // 1 instruction
    long long execints = 0; // 1 instruction
    int n; // 1 instruction
    execints += 1; // increment instruction count
    std::cin >> n; // 1 instruction
    execints += 1; // increment instruction count
    std::vector<int> numbers(n); // n + 2 instructions
    execints += n + 2; // increment instruction count

    for (int i = 0; i < n; ++i) { // 3 instructions + (n + 1) * 2 instructions
        execints += 3; // increment instruction count
        execints += 2; // increment instruction count for comparison
        std::cin >> numbers[i]; // n instructions
        execints += 1; // increment instruction count
    }

    auto result = MaxPairwiseProduct(numbers, execints); // 1 instruction
    execints += 1; // increment instruction count
    std::cout << "Number of executed instructions: " << execints << std::endl; // 1 instruction
    return 0; // 1 instruction
}


Overwriting max_pairwise_product_long_efficient_execints.cpp


Test with inputs 1 and 2



In [87]:
!g++ -o max_pairwise_product_long_efficient_execints max_pairwise_product_long_efficient_execints.cpp
!time ./max_pairwise_product_long_efficient_execints < input1.txt
!time ./max_pairwise_product_long_efficient_execints < input2.txt

Number of executed instructions: 147

real	0m0.004s
user	0m0.001s
sys	0m0.002s
Number of executed instructions: 130

real	0m0.003s
user	0m0.001s
sys	0m0.002s


Test and time it with input 3

In [88]:
!g++ -o max_pairwise_product_long_efficient_execints max_pairwise_product_long_efficient_execints.cpp
!time ./max_pairwise_product_long_efficient_execints < input3.txt


Number of executed instructions: 87

real	0m0.003s
user	0m0.003s
sys	0m0.000s


In [89]:
#Test input11
!g++ -o max_pairwise_product_long_efficient_execints max_pairwise_product_long_efficient_execints.cpp
!time ./max_pairwise_product_long_efficient_execints < input11.txt


Number of executed instructions: 2800007

real	0m0.075s
user	0m0.067s
sys	0m0.006s


20 stress tests

In [106]:
%%writefile run_all_test.sh
#!/bin/bash
g++ max_pairwise_product_long_efficient_execints.cpp -o max_pairwise_product_long_efficient_execints
for i in {13..32}
do
    ./max_pairwise_product_long_efficient_execints < input$i.txt > ctest$i.txt
    echo "Elapsed time on input$i"
    time ./max_pairwise_product_long_efficient_execints < input$i.txt
    echo ""
    grep -xvFf output$i.txt ctest$i.txt
done

Overwriting run_all_test.sh


In [107]:
!chmod u+x run_all_test.sh
!./run_all_test.sh

Elapsed time on input13
Number of executed instructions: 9430

real	0m0.003s
user	0m0.001s
sys	0m0.002s

Number of executed instructions: 9430
Elapsed time on input14
Number of executed instructions: 8766

real	0m0.002s
user	0m0.000s
sys	0m0.002s

Number of executed instructions: 8766
Elapsed time on input15
Number of executed instructions: 4826

real	0m0.002s
user	0m0.001s
sys	0m0.001s

Number of executed instructions: 4826
Elapsed time on input16
Number of executed instructions: 3435

real	0m0.002s
user	0m0.001s
sys	0m0.001s

Number of executed instructions: 3435
Elapsed time on input17
Number of executed instructions: 7710

real	0m0.002s
user	0m0.001s
sys	0m0.001s

Number of executed instructions: 7710
Elapsed time on input18
Number of executed instructions: 5610

real	0m0.002s
user	0m0.000s
sys	0m0.002s

Number of executed instructions: 5610
Elapsed time on input19
Number of executed instructions: 1688

real	0m0.002s
user	0m0.002s
sys	0m0.000s

Number of executed instructions: 1688

Count executed instructions in code and verify with inputs 1,2 and 3


In [92]:
!g++ -o max_pairwise_product_long_efficient_execints max_pairwise_product_long_efficient_execints.cpp
!./max_pairwise_product_long_efficient_execints < input1.txt
!./max_pairwise_product_long_efficient_execints < input2.txt
!./max_pairwise_product_long_efficient_execints < input3.txt


Number of executed instructions: 147
Number of executed instructions: 130
Number of executed instructions: 87


Calculate estimated time by instruction and number of instructions by second

In [108]:
#Compute the time using some of the input files
!g++ max_pairwise_product_long_sorting_execints.cpp -o max_pairwise_product_long_sorting_execints
!time ./max_pairwise_product_long_sorting_execints < input12.txt  > output12.txt
!time ./max_pairwise_product_long_sorting_execints < input27.txt  > output27.txt
!time ./max_pairwise_product_long_sorting_execints < input31.txt  > output31.txt



real	0m0.068s
user	0m0.060s
sys	0m0.006s

real	0m0.005s
user	0m0.003s
sys	0m0.000s

real	0m0.003s
user	0m0.001s
sys	0m0.002s


In [95]:
# Read the content of output files
with open('output12.txt', 'r') as file:
    output1_content = file.read().strip()

with open('output27.txt', 'r') as file:
    output2_content = file.read().strip()

with open('output31.txt', 'r') as file:
    output3_content = file.read().strip()

# Parse the number of executed instructions
executed_instructions_1 = int(output1_content)
executed_instructions_2 = int(output2_content)
executed_instructions_3 = int(output3_content)

# Real-time measurements in seconds
real_time_1 = 0.066
real_time_2 = 0.004
real_time_3 = 0.005

# Compute Estimated Time per Instruction (ETI) and Instructions per Second (IPS)
eti_1 = real_time_1 / executed_instructions_1
ips_1 = executed_instructions_1 / real_time_1

eti_2 = real_time_2 / executed_instructions_2
ips_2 = executed_instructions_2 / real_time_2

eti_3 = real_time_3 / executed_instructions_3
ips_3 = executed_instructions_3 / real_time_3

# Print the results
print("Output 1:")
print("ETI: {:.9f} s/instruction".format(eti_1))
print("IPS: {:.0f} instructions/second".format(ips_1))

print("\nOutput 2:")
print("ETI: {:.9f} s/instruction".format(eti_2))
print("IPS: {:.0f} instructions/second".format(ips_2))

print("\nOutput 3:")
print("ETI: {:.9f} s/instruction".format(eti_3))
print("IPS: {:.0f} instructions/second".format(ips_3))



Output 1:
ETI: 0.000000165 s/instruction
IPS: 6060621 instructions/second

Output 2:
ETI: 0.000004334 s/instruction
IPS: 230750 instructions/second

Output 3:
ETI: 0.000015480 s/instruction
IPS: 64600 instructions/second


### Estimate and verify the maximun n that can be solved in 1 second

In [1]:
# Estimate n with efficient solution

%%writefile max_n_in_a_second_long_efficient.cpp
#include <iostream>
#include <vector>
#include <chrono>
#include <fstream>
#include <random>
#include <limits>

// Function to generate a random integer in a specified range
int random_int(int min, int max) {
    static std::random_device rd;
    static std::mt19937 gen(rd());
    std::uniform_int_distribution<int> dis(min, max);
    return dis(gen);
}

// Function to generate input data of size n
void generate_input_data(int n, int min_value, int max_value) {
    std::ofstream input_file("input.txt");
    input_file << n << "\n"; // Output vector size

    for (int j = 0; j < n; ++j) {
        int num = random_int(min_value, max_value); // Generate random values for vector
        input_file << num << " "; // Output vector element
    }
    input_file << "\n"; // End of vector
    input_file.close();
}

std::pair<long long, long long> MaxPairwiseProduct(const std::vector<int>& numbers) {
    long long max_product = 0;
    long long second_max = 0;

    for (int i = 0; i < numbers.size(); ++i) {
        if (numbers[i] > max_product) {
            second_max = max_product;
            max_product = numbers[i];
        } else if (numbers[i] > second_max) {
            second_max = numbers[i];
        }
    }

    return std::make_pair(max_product, second_max);
}

int main() {
    int n = 10;
    int min_value = 1; // Minimum value for random numbers
    int max_value = 1000000; // Maximum value for random numbers
    int max_n = 1000000;  // Maximum value of n to try
    auto start_time = std::chrono::high_resolution_clock::now();
    while (n <= max_n) {
        generate_input_data(n, min_value, max_value); // Generate input data for current n
        auto start = std::chrono::high_resolution_clock::now();
        // Read input data from file
        std::ifstream input_file("input.txt");
        int num;
        std::vector<int> input_data;
        input_file >> num;
        input_data.reserve(num);
        while (input_file >> num) {
            input_data.push_back(num);
        }
        input_file.close();
        MaxPairwiseProduct(input_data);
        auto end = std::chrono::high_resolution_clock::now();
        double elapsed_time = std::chrono::duration_cast<std::chrono::milliseconds>(end - start).count() / 1000.0;
        std::cout << "n = " << n << ", Time taken: " << elapsed_time << " seconds" << std::endl;
        if (elapsed_time >= 1.0) {
            break;
        }
        n *= 2;
    }

    std::cout << "Maximum n that can be solved in 1 second: " << n << std::endl;

    return 0;
}



Writing max_n_in_a_second_long_efficient.cpp


In [2]:
!g++ -o max_n_in_a_second_long_efficient max_n_in_a_second_long_efficient.cpp
!./max_n_in_a_second_long_efficient


n = 10, Time taken: 0 seconds
n = 20, Time taken: 0 seconds
n = 40, Time taken: 0 seconds
n = 80, Time taken: 0 seconds
n = 160, Time taken: 0 seconds
n = 320, Time taken: 0 seconds
n = 640, Time taken: 0 seconds
n = 1280, Time taken: 0 seconds
n = 2560, Time taken: 0 seconds
n = 5120, Time taken: 0.001 seconds
n = 10240, Time taken: 0.001 seconds
n = 20480, Time taken: 0.002 seconds
n = 40960, Time taken: 0.005 seconds
n = 81920, Time taken: 0.011 seconds
n = 163840, Time taken: 0.021 seconds
n = 327680, Time taken: 0.045 seconds
n = 655360, Time taken: 0.091 seconds
Maximum n that can be solved in 1 second: 1310720


## Super efficient solution

Without storing the list in memory, only stpring the largest and second largest and whitout calling a function, add in each line a coment with the number of instructions that are executed in that line

in C++ and Python


 $$T(n) < 13 n  = O(n)$$


In [98]:
%%writefile max_pairwise_product_super_efficient.cpp
#include <iostream>

int main() {
    int n;
    std::cin >> n;

    // Initialize variables to store the largest and second largest numbers
    long long largest = 0, second_largest = 0;

    for (int i = 0; i < n; ++i) {
        int number;
        std::cin >> number;

        // Update largest and second largest numbers
        if (number > largest) {
            second_largest = largest;
            largest = number;
        } else if (number > second_largest) {
            second_largest = number;
        }
    }

    // Calculate the maximum pairwise product
    long long max_product = largest * second_largest;
    std::cout << max_product << "\n";

    return 0;
}


Overwriting max_pairwise_product_super_efficient.cpp


Test with inputs 1 and 2




In [100]:
!g++ -o max_pairwise_product_super_efficient max_pairwise_product_super_efficient.cpp
!time ./max_pairwise_product_super_efficient < input1.txt
!time ./max_pairwise_product_super_efficient < input2.txt

90

real	0m0.004s
user	0m0.001s
sys	0m0.003s
90

real	0m0.005s
user	0m0.001s
sys	0m0.003s


Test and time it with input 3

In [102]:
!time ./max_pairwise_product_super_efficient < input3.txt

40000000000

real	0m0.005s
user	0m0.001s
sys	0m0.003s


In [103]:
#Test input13
!time ./max_pairwise_product_super_efficient < input13.txt

99560340

real	0m0.004s
user	0m0.001s
sys	0m0.003s


20 stress tests

In [104]:
%%writefile run_all_test.sh
#!/bin/bash
g++ max_pairwise_product_super_efficient.cpp -o max_pairwise_product_super_efficient
for i in {13..32}
do
    ./max_pairwise_product_full < input$i.txt > ctest$i.txt
    echo "Elapsed time on input$i"
    time ./max_pairwise_product_full < input$i.txt
    echo ""
    grep -xvFf output$i.txt ctest$i.txt
done

Overwriting run_all_test.sh


In [105]:
!chmod u+x run_all_test.sh
!./run_all_test.sh

Elapsed time on input13
99560340

real	0m0.007s
user	0m0.004s
sys	0m0.002s

Elapsed time on input14
99550500

real	0m0.007s
user	0m0.005s
sys	0m0.002s

Elapsed time on input15
98922835

real	0m0.003s
user	0m0.002s
sys	0m0.001s

Elapsed time on input16
99510220

real	0m0.003s
user	0m0.000s
sys	0m0.003s

Elapsed time on input17
99770126

real	0m0.005s
user	0m0.005s
sys	0m0.000s

Elapsed time on input18
99780040

real	0m0.004s
user	0m0.003s
sys	0m0.001s

Elapsed time on input19
95411457

real	0m0.002s
user	0m0.001s
sys	0m0.001s

Elapsed time on input20
98316270

real	0m0.003s
user	0m0.002s
sys	0m0.001s

Elapsed time on input21
99340765

real	0m0.005s
user	0m0.003s
sys	0m0.002s

Elapsed time on input22
99311190

real	0m0.010s
user	0m0.009s
sys	0m0.001s

Elapsed time on input23
99660264

real	0m0.005s
user	0m0.005s
sys	0m0.000s

Elapsed time on input24
99840039

real	0m0.009s
user	0m0.008s
sys	0m0.001s

Elapsed time on input25
99231176

real	0m0.004s
user	0m0.001s
sys	0m0.003s

99231176
Ela

In [109]:
%%writefile max_pairwise_product_super_efficient_execints.cpp
#include <iostream>

int main() {
    int n;
    std::cin >> n;

    // Initialize variables to store the largest and second largest numbers
    long long largest = 0, second_largest = 0;

    // Counter for the number of executed instructions
    long long execints = 0;

    for (int i = 0; i < n; ++i) {
        execints += 3; // Increment for variable initialization and loop condition check
        int number;
        std::cin >> number;
        execints += 2; // Increment for input operation

        // Update largest and second largest numbers
        execints += 4; // Increment for comparisons and assignments inside the loop
        if (number > largest) {
            second_largest = largest;
            largest = number;
            execints += 2; // Increment for assignments
        } else if (number > second_largest) {
            second_largest = number;
            execints += 1; // Increment for assignment
        }
    }

    // Calculate the maximum pairwise product
    long long max_product = largest * second_largest;
    execints += 1; // Increment for multiplication
    std::cout << "Maximum pairwise product: " << max_product << "\n";
    execints += 2; // Increment for output operation and newline character

    // Output the number of executed instructions
    std::cout << "Number of executed instructions: " << execints << std::endl;
    execints += 2; // Increment for output operation and newline character

    return 0;
}


Writing max_pairwise_product_super_efficient_execints.cpp


Count executed instructions in code and verify with inputs 1,2 and 3

In [116]:
!g++ -o max_pairwise_product_super_efficient_execints max_pairwise_product_super_efficient_execints.cpp
!./max_pairwise_product_super_efficient_execints < input1.txt
!./max_pairwise_product_super_efficient_execints < input2.txt
!./max_pairwise_product_super_efficient_execints < input3.txt

Maximum pairwise product: 90
Number of executed instructions: 113
Maximum pairwise product: 90
Number of executed instructions: 96
Maximum pairwise product: 40000000000
Number of executed instructions: 65


Calculate estimated time by instruction and number of instructions by second

In [118]:
#Compute the time using some of the input files
!g++ max_pairwise_product_super_efficient_execints.cpp -o max_pairwise_product_super_efficient_execints
!time ./max_pairwise_product_super_efficient_execints < input12.txt  > output12.txt
!time ./max_pairwise_product_super_efficient_execints < input27.txt  > output27.txt
!time ./max_pairwise_product_super_efficient_execints < input31.txt  > output31.txt



real	0m0.070s
user	0m0.067s
sys	0m0.002s

real	0m0.003s
user	0m0.001s
sys	0m0.002s

real	0m0.005s
user	0m0.002s
sys	0m0.002s


### Super efficient solution in Python

Without storing the list in memory and without calling a function, add in each line a comment with the number of instruction that are executed in that line

Test with inputs 1 and 2




In [ ]:
# Estimate n with "super efficient implementation"
%%writefile max_n_in_a_second_super_efficient.cpp

#include <iostream>
#include <vector>
#include <chrono>
#include <fstream>
#include <random>
#include <limits>

// Function to generate a random integer in a specified range
int random_int(int min, int max) {
    static std::random_device rd;
    static std::mt19937 gen(rd());
    std::uniform_int_distribution<int> dis(min, max);
    return dis(gen);
}

// Function to generate input data of size n
void generate_input_data(int n, int min_value, int max_value) {
    std::ofstream input_file("input.txt");
    input_file << n << "\n"; // Output vector size

    for (int j = 0; j < n; ++j) {
        int num = random_int(min_value, max_value); // Generate random values for vector
        input_file << num << " "; // Output vector element
    }
    input_file << "\n"; // End of vector
    input_file.close();
}

long long super_efficient_max_pairwise_product(const std::vector<int>& numbers) {
    long long largest = std::numeric_limits<long long>::min();
    long long second_largest = std::numeric_limits<long long>::min();

    for (int num : numbers) {
        if (num > largest) {
            second_largest = largest;
            largest = num;
        } else if (num > second_largest) {
            second_largest = num;
        }
    }

    return largest * second_largest;
}

int main() {
    int n = 10;
    int min_value = 1; // Minimum value for random numbers
    int max_value = 1000000; // Maximum value for random numbers
    int max_n = 1000000;  // Maximum value of n to try
    auto start_time = std::chrono::high_resolution_clock::now();
    while (n <= max_n) {
        generate_input_data(n, min_value, max_value); // Generate input data for current n
        auto start = std::chrono::high_resolution_clock::now();
        // Read input data from file
        std::ifstream input_file("input.txt");
        int num;
        std::vector<int> input_data;
        input_file >> num;
        input_data.reserve(num);
        while (input_file >> num) {
            input_data.push_back(num);
        }
        input_file.close();
        super_efficient_max_pairwise_product(input_data);
        auto end = std::chrono::high_resolution_clock::now();
        double elapsed_time = std::chrono::duration_cast<std::chrono::milliseconds>(end - start).count() / 1000.0;
        std::cout << "n = " << n << ", Time taken: " << elapsed_time << " seconds" << std::endl;
        if (elapsed_time >= 1.0) {
            break;
        }
        n *= 2;
    }

    std::cout << "Maximum n that can be solved in 1 second: " << n << std::endl;

    return 0;
}


Overwriting max_n_in_a_second_super_efficient.cpp


In [ ]:
!g++ -o max_n_in_a_second_super_efficient max_n_in_a_second_super_efficient.cpp
!./max_n_in_a_second_super_efficient


n = 10, Time taken: 0 seconds
n = 20, Time taken: 0 seconds
n = 40, Time taken: 0 seconds
n = 80, Time taken: 0 seconds
n = 160, Time taken: 0 seconds
n = 320, Time taken: 0 seconds
n = 640, Time taken: 0 seconds
n = 1280, Time taken: 0 seconds
n = 2560, Time taken: 0 seconds
n = 5120, Time taken: 0.001 seconds
n = 10240, Time taken: 0.001 seconds
n = 20480, Time taken: 0.003 seconds
n = 40960, Time taken: 0.005 seconds
n = 81920, Time taken: 0.011 seconds
n = 163840, Time taken: 0.023 seconds
n = 327680, Time taken: 0.043 seconds
n = 655360, Time taken: 0.096 seconds
Maximum n that can be solved in 1 second: 1310720


In [ ]:
#Create stress test of size n = 1310720
!g++ -o stress_test_of_size_n stress_test_of_size_n.cpp
!./stress_test_of_size_n

Enter the size of the vector (n): ^C


In [ ]:
#Verify the max_pairwise_product_super_efficient can run stress test of size n = 1310720

!g++ -o max_pairwise_product_efficient max_pairwise_product_efficient.cpp
!time ./max_pairwise_product_efficient < stress_test_of_size_n.txt

40000000000

real	0m1.093s
user	0m1.069s
sys	0m0.015s


## Perform 100 randomized automated tests in C++ and Python with a program that  compares the answers, for the same 100 random input instances, of the super efficient solution $(T(n) < 10 n = O(n))$ against the sorting solution  $(T(n)  = O(n \lg{n}))$,  if the answers are not the same stops and displays "Error" listing  the input and the two answers; if the 100 tests are completed without stoping displays "Pass".

In [137]:
%%writefile super_efficient_vs_sorting.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <random>

// Function to generate random input vector
std::vector<int> generate_random_input(int n, int min_value, int max_value) {
    std::vector<int> input;
    std::random_device rd;
    std::mt19937 gen(rd());
    std::uniform_int_distribution<int> dis(min_value, max_value);

    for (int i = 0; i < n; ++i) {
        input.push_back(dis(gen));
    }

    return input;
}

// Super-efficient solution
long long super_efficient_max_pairwise_product(const std::vector<int>& numbers) {
    long long largest = 0, second_largest = 0;

    for (int num : numbers) {
        if (num > largest) {
            second_largest = largest;
            largest = num;
        } else if (num > second_largest) {
            second_largest = num;
        }
    }

    return largest * second_largest;
}

// Sorting solution
long long sort_max_pairwise_product(const std::vector<int>& numbers) {
    std::vector<int> sorted_numbers = numbers;
    std::sort(sorted_numbers.begin(), sorted_numbers.end());
    int n = sorted_numbers.size();
    return (long long)sorted_numbers[n - 1] * sorted_numbers[n - 2];
}

int main() {
    const int num_tests = 100;
    const int n = 10; // Change this to the desired input size
    const int min_value = 1; // Minimum value for random numbers
    const int max_value = 1000; // Maximum value for random numbers

    bool error = false;

    for (int i = 0; i < num_tests; ++i) {
        // Generate random input
        std::vector<int> input = generate_random_input(n, min_value, max_value);

        // Calculate outputs for both solutions
        long long super_efficient_result = super_efficient_max_pairwise_product(input);
        long long sort_result = sort_max_pairwise_product(input);

        // Compare the outputs
        if (super_efficient_result != sort_result) {
            std::cout << "Error: Outputs mismatch for test " << i + 1 << std::endl;
            std::cout << "Input: ";
            for (int num : input) {
                std::cout << num << " ";
            }
            std::cout << std::endl;
            std::cout << "Super-efficient result: " << super_efficient_result << std::endl;
            std::cout << "Sorting result: " << sort_result << std::endl;
            error = true;
            break;
        }
    }

    if (!error) {
        std::cout << "Pass: All tests passed successfully." << std::endl;
    }

    return 0;
}



Overwriting super_efficient_vs_sorting.cpp


In [120]:
#Create 100 random stress tests

!g++ stress_test_generator.cpp -o stress_test_generator
!./stress_test_generator

#Start index: 101 for file100
#numbers of tests: 100 (from 101 to 200)


Enter the start index for input files: 101
Enter the number of stress tests to create: 100


In [144]:
%%writefile run_all_test.sh
#!/bin/bash

# Compile the C++ program
g++ super_efficient_vs_sorting.cpp -o super_efficient_vs_sorting

# Loop through input files from input13.txt to input32.txt
for i in {101..200}
do
    # Run the C++ program with input$i.txt and redirect output to ctest$i.txt
    ./super_efficient_vs_sorting < input$i.txt > ctest$i.txt

    # Display elapsed time for the current input file
    echo "Comparison on input$i"
    #time ./super_efficient_vs_sorting < input$i.txt
    echo ""

    # Compare the output of the program with the expected output
    grep -xvFf output$i.txt ctest$i.txt
done


Overwriting run_all_test.sh


In [143]:
!chmod u+x run_all_test.sh
!./run_all_test.sh

Comparison on input1

Pass: All tests passed successfully.
Comparison on input2

Pass: All tests passed successfully.
Comparison on input3

Pass: All tests passed successfully.
Comparison on input4

Pass: All tests passed successfully.
Comparison on input5

Pass: All tests passed successfully.
Comparison on input6

Pass: All tests passed successfully.
Comparison on input7

Pass: All tests passed successfully.
Comparison on input8

Pass: All tests passed successfully.
Comparison on input9

Pass: All tests passed successfully.
Comparison on input10

Pass: All tests passed successfully.
Comparison on input11

Pass: All tests passed successfully.
Comparison on input12

Pass: All tests passed successfully.
Comparison on input13

Pass: All tests passed successfully.
Comparison on input14

Pass: All tests passed successfully.
Comparison on input15

Pass: All tests passed successfully.
Comparison on input16

Pass: All tests passed successfully.
Comparison on input17

Pass: All tests passed suc